src/aggregation/bucket/range.rs

```diff
@@ -0,0 +1,546 @@
+        range.from..range.to
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeAndBucketEntry {
+    range: Range<u64>,
+    bucket: SegmentRangeBucketEntry,
+}
+
+/// The collector puts values from the fast field into the correct buckets and does a conversion to
+/// the correct datatype.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeCollector {
+    /// The buckets containing the aggregation data.
+    buckets: Vec<SegmentRangeAndBucketEntry>,
+    field_type: Type,
+}
+
+impl SegmentRangeCollector {
+    pub fn into_intermediate_bucket_result(self) -> IntermediateBucketResult {
+        let field_type = self.field_type;
+
+        let buckets = self
+            .buckets
+            .into_iter()
+            .map(move |range_bucket| {
+                (
+                    range_to_key(&range_bucket.range, &field_type),
+                    range_bucket.bucket.into(),
+                )
+            })
+            .collect();
+
+        IntermediateBucketResult::Range(buckets)
+    }
+
+    pub(crate) fn from_req(
+        req: &RangeAggregation,
+        sub_aggregation: &AggregationsWithAccessor,
+        field_type: Type,
+    ) -> crate::Result<Self> {
+        // The range input on the request is f64.
+        // We need to convert to u64 ranges, because we read the values as u64.
+        // The mapping from the conversion is monotonic so ordering is preserved.
+        let buckets = extend_validate_ranges(&req.ranges, &field_type)?
+            .iter()
+            .map(|range| {
+                let to = if range.end == u64::MAX {
+                    None
+                } else {
+                    Some(f64_from_fastfield_u64(range.end, &field_type))
+                };
+                let from = if range.start == u64::MIN {
+                    None
+                } else {
+                    Some(f64_from_fastfield_u64(range.start, &field_type))
+                };
+                Ok(SegmentRangeAndBucketEntry {
+                    range: range.clone(),
+                    bucket: SegmentRangeBucketEntry {
+                        key: range_to_key(range, &field_type),
+                        doc_count: 0,
+                        sub_aggregation: SegmentAggregationResultsCollector::from_req(
+                            sub_aggregation,
+                        )?,
+                        from,
+                        to,
+                    },
+                })
+            })
+            .collect::<crate::Result<_>>()?;
+
+        Ok(SegmentRangeCollector {
+            buckets,
+            field_type,
+        })
+    }
+
+    #[inline]
+    pub(crate) fn collect_block(
+        &mut self,
+        doc: &[DocId],
+        bucket_with_accessor: &BucketAggregationWithAccessor,
+        force_flush: bool,
+    ) {
+        let mut iter = doc.chunks_exact(4);
+        for docs in iter.by_ref() {
+            let val1 = bucket_with_accessor.accessor.get(docs[0]);
```

Comment:
why all the manual unrolling?

Reply:
It's faster

```
test aggregation::tests::bench::bench_aggregation_average_f64                                                            ... bench:   7,063,803 ns/iter (+/- 1,769,966)
test aggregation::tests::bench::bench_aggregation_average_u64                                                            ... bench:   5,622,493 ns/iter (+/- 100,153)
test aggregation::tests::bench::bench_aggregation_average_u64_and_f64                                                    ... bench:   8,695,562 ns/iter (+/- 109,397)
test aggregation::tests::bench::bench_aggregation_stats_f64                                                              ... bench:   7,493,152 ns/iter (+/- 214,351)
test aggregation::tests::bench::bench_aggregation_sub_tree                                                               ... bench:  13,159,764 ns/iter (+/- 589,236)
```

removed unrolling in average metric
```
test aggregation::tests::bench::bench_aggregation_average_f64                                                            ... bench:   7,631,327 ns/iter (+/- 174,053)
test aggregation::tests::bench::bench_aggregation_average_u64                                                            ... bench:   7,451,904 ns/iter (+/- 410,011)
test aggregation::tests::bench::bench_aggregation_average_u64_and_f64                                                    ... bench:  10,471,694 ns/iter (+/- 140,192)
test aggregation::tests::bench::bench_aggregation_stats_f64                                                              ... bench:   7,465,887 ns/iter (+/- 344,009)
test aggregation::tests::bench::bench_aggregation_sub_tree                                                               ... bench:  16,609,919 ns/iter (+/- 399,840)
```

Reply:
Awesome ! Thanks